### PR TITLE
[PyTorch/Jax] Fix attention mask definition, and sliding window for decoder

### LINKED ISSUE #614: [PyTorch] Definition of attention mask
Hi team,

I am wondering about the definition of the attention mask in transformer-engine. I did not find an explanation in the docs. Does True mean that the position takes part in attention or that it is masked out?

These two code positions suggest that True means masked out:
https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/test_numerics.py#L71
https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/utils.py#L34

But for flash attention with padding, the cumulated sequence lengths are computed based on a sum over the mask. This suggests that it is vice versa:
https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention.py#L180

Thanks for clarifying this!

### LINKED ISSUE #629: [PyTorch] FlashAttention: causal masking enforced in cross attention due to sliding window attention
Hi transformer-engine team,

we noticed that in a decoder layer with `self_attn_mask_type="causal"`, the following line https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/transformer.py#L594 sets window_size = (-1, 0). This window_size is passed to both self attention as well as to cross attention. In the cross attention this encforces a causal masking although padding is intended. This bug was probably introduced in https://github.com/NVIDIA/TransformerEngine/pull/551.

I am unsure what a poper fix might be. Does a windowed attention really make sense for cross attention with padding? Considering that the FlashAttention backend re-packs the sequences with `PackTensors`, this might have unintended side effects.

Cheers
Markus

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

Fixes #614.
Fixes #629.

Moving forward, we'd like to define attention mask consistently in PyTorch, Jax and Paddle as `True` being masking out the corresponding position and `False` being allowing that position to participate in attention. A typical causal mask would look like:
```
0, 1, 1, 1
0, 0, 1, 1
0, 0, 0, 1
0, 0, 0, 0
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

- Unify the implementation and unit tests regarding attention mask in TE PyTorch, Jax and Paddle.
- Differentiate sliding widow initialization for decoder and encoder
- Save `aux_ctx_tensors` and similar tensors using `save_for_backward` call instead of to `ctx` 
- Skip Jax tests for fused attention backward when bias shape is not 1HSS

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by timmoon10** [comment]
For reference, most frameworks treat attention mask `True`s as inclusion:

| Implementation | `True` in attention mask implies inclusion |
|----|----|
| [`torch.nn.MultiheadAttention`](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) | No |
| [`torch.nn.functional.scaled_dot_product_attention`](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch.nn.functional.scaled_dot_product_attention) | Yes |
| [`flax.linen.MultiHeadAttention`](https://flax.readthedocs.io/en/latest/_modules/flax/linen/attention.html) | Yes |
| [`paddle.nn.MultiHeadAttention`](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/MultiHeadAttention_en.html) | Yes |
| [`keras.layers.Attention`](https://keras.io/api/layers/attention_layers/attention/) | Yes |
| [`tf.keras.layers.Attention`](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Attention) | Yes |
| [cuDNN fused attention](https://docs.nvidia.com/deeplearning/cudnn/latest/developer/graph-api.html#fused-flash-attention-fprop) | Yes |
| [`apex.contrib.multihead_attn.SelfMultiheadAttn`](https://github.com/NVIDIA/apex/blob/master/apex/contrib/multihead_attn/self_multihead_attn.py) | No |
| [`flash_attn.fused_softmax.ScaledMaskedSoftmax`](https://github.com/Dao-AILab/flash-attention/blob/9a11f440d3a34f618b4ba814c825b109c6d7e8f5/flash_attn/fused_softmax.py#L73) | No |

**Comment by cyanguwa** [comment]
@timmoon10 thanks for the table :smiley_cat: I was doing something like that myself (OCD? not if someone else is also doing it!).

After looking through some of the other frameworks' implementations (PyTorch/Jax/Paddle), and our own implementation and unit tests, I'm leaning towards `True = masking out; False = inclusion in attention` actually. 

I don't think Keras is a popular framework now and that we should follow its trend. Also, PyTorch's `scaled_dot_product_attention` is a beta feature but its `MultiHeadAttention` is stable. cuDNN's attention doesn't really use a mask in my opinion, not the `True/False` kind anyway. For causal, it's using a `is_causal=True/False` flag, and for padding, a `cu_seqlens` tensor (which we converted the mask to in TE).

Our fused softmax implementation uses `True = masking out; False = inclusion in attention`, which also makes all the softmax calls in TE PyTorch, Jax and Paddle all subscribe to that definition. They are not the fastest attention paths, so maybe we don't have to care about them as much. But, I think the original term when it came out, for causal, is `upper triangle masking`, which kind of made it `upper` rather than `lower`, and that 1 means masking things out. This is probably also why apex and Tri Dao attention's softmax both adopted that definition.

Any thoughts? :smiley: thanks!

**Comment by zlsh80826** [comment]
Hi @cyanguwa, thanks for making things clear. TE-JAX awares mask_bit = `True` to mask out, and it is [documented](https://github.com/NVIDIA/TransformerEngine/blob/8e672ff0758033c348e263dbcd6a4b3578c01161/transformer_engine/jax/flax/transformer.py#L1481), so I think JAX's unit tests don't need to be changed in this PR.

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -130,10 +130,7 @@ def customcall_fused_dpa(query, key, value, bias, q_token, kv_token, dropout_rng
     if is_causal_mask(attn_mask_type):
         mask = make_decoder_mask(q_token, kv_token)
     else:
-        mask = make_attention_mask(q_token > 0, kv_token > 0)
-
-    # mask invert
-    mask = jnp.logical_not(mask)
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:136]: `mask = jnp.logical_not(mask)` is to make sure the `mask_bit = True` means mask out, so the original code already covers this PR's definition.

Comment (cyanguwa) [reply tests/jax/test_fused_attn.py:136]: I appreciate that Jax has used the `True-mask out; False-keep` definition of attention mask and documented it. I also understand that the logic here can achieve whatever conversion is needed and make the tests pass. But I don't want users to accidentally borrow some of the mask generation functions here thinking that we have defined the mask the other way. Issue 614 was raised because of some consistency between implementation and tests in PyTorch, so I wanted to avoid that here.

Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:136]: The conversion here is not just to "make the tests pass", it is what exactly users need to do w/ and w/o TE in the JAX eco system. In JAX's eco system, `False-mask out; True-keep` is widely used across different libraries ([flax](https://github.com/google/flax/blob/40c7eaa74817a1716d3f64ad70a292e975837753/flax/linen/attention.py#L752), [t5x](https://github.com/search?q=repo%3Agoogle-research%2Ft5x%20make_decoder_mask&type=code)), so I hope the `make_xxx_mask` just generate the `False-mask out; True-keep` to align the existing conventions. For TE's user, they don't need to change any implementation details in `make_xxx_mask` or maintaining a TE version's mask, what they need to do is
```
if use_te:
    mask = jnp.logical_not(mask)
```
I still think the original tests is the most clear way to demonstrate how to use the TE's fused attention API.

**Comment by Marks101** [comment]
@cyanguwa Thank you so much for looking into this and clarifying the mask definition ❤️

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (cyanguwa) [reply tests/jax/test_fused_attn.py:136]: I understand your point and I did notice that Flax uses a different definition (`True-keep; False-mask out`) in its `[flax.linen.make_causal_mask](https://flax.readthedocs.io/en/v0.5.3/_autosummary/flax.linen.make_causal_mask.html)`. I think Flax carries some weight here, but t5x maybe not so much - it's an important model but probably not a library like Flax is.

On the flip side, the purpose of this PR was to make TE consistent - across all three frameworks, and across implementation code and test code. We would want users to just see one definition when they see a mask tensor. 

@ptrendx any preference here?

Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:136]: I've got your point of concern, I think we were focuing on the different parts of the code. I agree we can change the implementation of reference code and I was focusing on how to generate the mask clearly.

How about this:

- Keep your change for `logits = jnp.where(mask, jnp.finfo(dtype).min, logits)`, which aligns the TE's code and reference code.
- For mask generation
  - Keep `make_decoder_mask` as the original
  - Add a function `make_mask` 
  ```
  def make_mask(q_token, kv_token, attn_mask_type)
     """Create True-mask out; False-keep mask"""
     if is_causal_mask(attn_mask_type):
       inv_mask = make_decoder_mask(q_token, kv_token)
     else:
       inv_mask = make_attention_mask(q_token > 0, kv_token > 0)
     mask = jnp.logical_not(inv_mask)
     return mask
  ```
  - Change the mask generation in `customcall_fused_dpa ` and `jax_dpa` to `make_mask`.

Comment (cyanguwa) [reply tests/jax/test_fused_attn.py:136]: Done! 👍

**Review by zlsh80826 — APPROVED** [review]
LGTM!

**Review by ptrendx — APPROVED** [review]
LGTM
